meshio/msh_io/msh4_1.py

```diff
@@ -0,0 +1,452 @@
+                    "(got %s). Converting.",
+                    value.dtype,
+                )
+                mesh.cells[key] = numpy.array(value, dtype=c_int)
+
+    # Gmsh cells are mostly ordered like VTK, with a few exceptions:
+    cells = mesh.cells.copy()
+    if "tetra10" in cells:
+        cells["tetra10"] = cells["tetra10"][:, [0, 1, 2, 3, 4, 5, 6, 7, 9, 8]]
+    if "hexahedron20" in cells:
+        cells["hexahedron20"] = cells["hexahedron20"][
+            :, [0, 1, 2, 3, 4, 5, 6, 7, 8, 11, 13, 9, 16, 18, 19, 17, 10, 12, 14, 15]
+        ]
+
+    with open(filename, "wb") as fh:
+        mode_idx = 1 if write_binary else 0
+        size_of_double = 8
+        fh.write(
+            ("$MeshFormat\n4 {} {}\n".format(mode_idx, size_of_double)).encode("utf-8")
+        )
+        if write_binary:
+            fh.write(struct.pack("i", 1))
+            fh.write("\n".encode("utf-8"))
+        fh.write("$EndMeshFormat\n".encode("utf-8"))
+
+        if mesh.field_data:
+            _write_physical_names(fh, mesh.field_data)
+
+        _write_nodes(fh, mesh.points, write_binary)
+        _write_elements(fh, cells, write_binary)
+        if mesh.gmsh_periodic is not None:
+            _write_periodic(fh, mesh.gmsh_periodic)
+        for name, dat in mesh.point_data.items():
+            _write_data(fh, "NodeData", name, dat, write_binary)
+        cell_data_raw = raw_from_cell_data(mesh.cell_data)
+        for name, dat in cell_data_raw.items():
+            _write_data(fh, "ElementData", name, dat, write_binary)
+
+    return
+
+
+def _write_nodes(fh, points, write_binary):
+    fh.write("$Nodes\n".encode("utf-8"))
+
+    # TODO not sure what dimEntity is supposed to say
+    dim_entity = 0
+    type_node = 0
+
+    if write_binary:
+        # write all points as one big block
+        # numEntityBlocks(unsigned long) numNodes(unsigned long)
+        # tagEntity(int) dimEntity(int) typeNode(int) numNodes(unsigned long)
+        # tag(int) x(double) y(double) z(double)
+        fh.write(numpy.array([1, points.shape[0]], dtype=c_ulong).tostring())
+        fh.write(numpy.array([1, dim_entity, type_node], dtype=c_int).tostring())
+        fh.write(numpy.array([points.shape[0]], dtype=c_ulong).tostring())
+        dtype = [("index", c_int), ("x", c_double, (3,))]
+        tmp = numpy.empty(len(points), dtype=dtype)
+        tmp["index"] = 1 + numpy.arange(len(points))
+        tmp["x"] = points
+        fh.write(tmp.tostring())
+        fh.write("\n".encode("utf-8"))
+    else:
+        # write all points as one big block
+        # numEntityBlocks(unsigned long) numNodes(unsigned long)
+        fh.write("{} {}\n".format(1, len(points)).encode("utf-8"))
+
+        # tagEntity(int) dimEntity(int) typeNode(int) numNodes(unsigned long)
+        fh.write(
+            "{} {} {} {}\n".format(1, dim_entity, type_node, len(points)).encode(
+                "utf-8"
+            )
+        )
+
+        for k, x in enumerate(points):
```

Comment:
Is this still correct, i.e., are the tags still prepended to each line? We could think about removing this loop, too.

Reply:
This write method is plain wrong, it's just the meagre old MSH4.0 one, itself a quick update of the MSH2 method. I was going to defer writing MSH4.1.

Reply:
No, it's wrong, syntax for MSH4.1 is (at least for `parameter == 0`)
> nodeTag(size_t)
>    ...
>   x(double) y(double) z(double)

I had been hoping to defer `msh4_1.write` #332 but will get onto this now.

Reply:
As reported, #332 is stalled.